lualib/resty/kong/var.lua

```diff
@@ -108,7 +108,8 @@ local function var_get_by_index(index)
                                                         value_len, errmsg)
 
     if rc == NGX_OK then
-        return ffi_str(value_ptr[0], value_len[0])
```

Comment:
Does this trigger a NYI for tail call?

From the LuaJIT's tail call implementation, the tail call NYI only happen for variable argument function, but ffi_str seems not a variable argument function. 

https://github.com/LuaJIT/LuaJIT/blob/v2.1/src/lj_record.c#L858-L861

Reply:
If zhongweiy's comment is correct, should we close this PR?

Reply:
``` plain
---- TRACE 1322 start openresty/lualib/resty/kong/var.lua:181
0001  UGET     2   0      ; variable_index       (openresty/lualib/resty/kong/var.lua:182)
0002  TGETV    2   2   1       (openresty/lualib/resty/kong/var.lua:182)
0003  ISF          2       (openresty/lualib/resty/kong/var.lua:183)
0004  JMP      3 => 0008
0005  UGET     3   1      ; var_get_by_index       (openresty/lualib/resty/kong/var.lua:184)
0006  MOV      5   2       (openresty/lualib/resty/kong/var.lua:184)
0007  CALLT    3   2       (openresty/lualib/resty/kong/var.lua:184)
0000  FUNCF   10          ; openresty/lualib/resty/kong/var.lua:99
0001  UGET     1   0      ; get_request       (openresty/lualib/resty/kong/var.lua:100)
0002  CALL     1   2   1       (openresty/lualib/resty/kong/var.lua:100)
0000  . JFUNCF   4 1321         ; openresty/lualib/resty/kong/var.lua:60
0001  . UGET     0   0      ; orig_get_request       (openresty/lualib/resty/kong/var.lua:61)
0002  . CALL     0   2   1       (openresty/lualib/resty/kong/var.lua:61)
0000  . . FUNCF    2          ; openresty/lualib/resty/core/base.lua:252
0001  . . UGET     0   0      ; exdata       (openresty/lualib/resty/core/base.lua:253)
0002  . . CALL     0   2   1       (openresty/lualib/resty/core/base.lua:253)
0000  . . . FUNCC               ; thread.exdata
0003  . . ISEQP    0   0       (openresty/lualib/resty/core/base.lua:254)
0004  . . JMP      1 => 0006
0000  . . . . FUNCC               ; ffi.meta.__eq
0005  . . RET1     0   2       (openresty/lualib/resty/core/base.lua:255)
0003  . IST          0       (openresty/lualib/resty/kong/var.lua:63)
0004  . JMP      1 => 0008
0008  . RET1     0   2       (openresty/lualib/resty/kong/var.lua:67)
0003  UGET     2   1      ; get_size_ptr       (openresty/lualib/resty/kong/var.lua:102)
0004  CALL     2   2   1       (openresty/lualib/resty/kong/var.lua:102)
0000  . JFUNCF   3 1293         ; openresty/lualib/resty/core/base.lua:181
0001  . UGET     0   0      ; size_ptr       (openresty/lualib/resty/core/base.lua:182)
0002  . IST          0       (openresty/lualib/resty/core/base.lua:182)
0003  . JMP      0 => 0008
0008  . UGET     0   0      ; size_ptr       (openresty/lualib/resty/core/base.lua:186)
0009  . RET1     0   2       (openresty/lualib/resty/core/base.lua:186)
0005  UGET     3   2      ; C       (openresty/lualib/resty/kong/var.lua:107)
0006  TGETS    3   3   0  ; "ngx_http_lua_kong_ffi_var_get_by_index"       (openresty/lualib/resty/kong/var.lua:107)
0000  . . FUNCC               ; ffi.clib.__index
0007  MOV      5   1       (openresty/lualib/resty/kong/var.lua:107)
0008  MOV      6   0       (openresty/lualib/resty/kong/var.lua:107)
0009  UGET     7   3      ; value_ptr       (openresty/lualib/resty/kong/var.lua:107)
0010  MOV      8   2       (openresty/lualib/resty/kong/var.lua:108)
0011  UGET     9   4      ; errmsg       (openresty/lualib/resty/kong/var.lua:108)
0012  CALL     3   2   6       (openresty/lualib/resty/kong/var.lua:107)
0000  . FUNCC               ; ffi.meta.__call
0013  UGET     4   5      ; NGX_OK       (openresty/lualib/resty/kong/var.lua:110)
0014  ISNEV    3   4       (openresty/lualib/resty/kong/var.lua:110)
0015  JMP      4 => 0021
0016  UGET     4   6      ; ffi_str       (openresty/lualib/resty/kong/var.lua:111)
0017  UGET     6   3      ; value_ptr       (openresty/lualib/resty/kong/var.lua:111)
0018  TGETB    6   6   0       (openresty/lualib/resty/kong/var.lua:111)
0000  . . FUNCC               ; ffi.meta.__index
0019  TGETB    7   2   0       (openresty/lualib/resty/kong/var.lua:111)
0000  . . FUNCC               ; ffi.meta.__index
0020  CALLT    4   3       (openresty/lualib/resty/kong/var.lua:111)
0000  FUNCC               ; ffi.string
---- TRACE 1322 abort @./kong/init.lua:1059 -- NYI: return to lower frame
```

https://github.com/Kong/lua-kong-nginx-module/blob/d1f8aaf32702353ebee6caaa4c44f8b84ed8202b/lualib/resty/kong/var.lua#L111

***

There are many cases cause the RETL abort
<img width="845" alt="" src="https://github.com/Kong/lua-kong-nginx-module/assets/44437200/e29eaa1b-77a0-438c-84f4-10064ebd81d8">

To be honest, I didn't debug it deeply, but once I remove that, this NYI doesn't happen anymore.

Reply:
OK, then it is probably caused by some `RET` instruction here instead of tail call like `ffi_str()`.

@ADD-SP is it easy to reproduce it? If yes, I can take a look at which pattern exactly causes this NYI and may help work out a tip for performance sensitive code style.

Reply:
@zhongweiy Reproducing is very easy, just adding a single route and hitting it by `wrk`.

Reply:
Thanks, I will add it to my TODO list and will keep you updated.